### PR TITLE
Feature/migrate iterable to contain in any order only values expectations to kotlin test

### LINKED ISSUE #2025: migrate IterableToContainInAnyOrderOnlyValuesExpectationsSpec to kotlin-test
*Platform* (all, jvm, js): all
*Extension* (none, kotlin 1.3): none

## Code related feature

We want to move away from Spek as test-runner as we consider this project dead and it makes more and more problems in conjunction with intellij and gradle. Hence, we migrate the tests in atrium-specs to kotlin-test.

Following the things you need to do:

see https://github.com/robstoll/atrium/issues/1963

## Your first contribution?
- Write a comment `I'll work on this` if you would like to take this issue over. 
  This way we get the chance to revise the description in case things have changed in the meantime, we might give you additional hints and we can assign the task to you, so that others do not start as well.
- See [Your first code contribution](https://github.com/robstoll/atrium/blob/main/.github/CONTRIBUTING.md#your-first-code-contribution) for guidelines.  
- Do not hesitate to ask questions here or to contact us via [Atrium's slack channel](https://kotlinlang.slack.com/team/U3DE1TXKP) if you need help
  ([Invite yourself](https://slack.kotlinlang.org/) in case you do not have an account yet).

### PR DESCRIPTION
**Pull request by TomerPacific**
Resolves #2025 

This pull request refactors and improves the test infrastructure for the "Iterable to contain in any order only values" expectations in both the fluent and infix APIs. The main change is the introduction of a new abstract base test class to consolidate and standardize the test logic, and the renaming and updating of test files to use this new base. This helps reduce code duplication and makes the test suite easier to maintain and extend.

**Test infrastructure improvements:**

* Introduced a new abstract base test class `AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest` containing all shared test logic for "in any order only values" expectations, including various edge cases and reporting option scenarios.
* Updated the fluent API test (`IterableToContainInAnyOrderOnlyValuesExpectationsTest`) and infix API test (`IterableToContainInAnyOrderOnlyValuesExpectationsTest`) to inherit from the new abstract base class, and renamed them from `*Spec` to `*Test` for consistency. 

**Test suite refactoring:**

* Updated references in builder spec objects to use the new abstract base class instead of the old spec class, ensuring all tests use the consolidated logic. 

**My main points of unclarity are :**
- What values I need to pass when using `toContainInAnyOrderOnlyValuesFun` (I'll point some examples with comments)
- Getting errors on `showOnlyFailingIfMoreExpectedElementsThan `and similar methods
______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/main/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        ))
+    )
+
+    @TestFactory
+    fun empty_collection_1_throws() {
+        nonNullableCases(
+            toContainInAnyOrderOnlyValuesSpec,
+            toContainInAnyOrderOnlyNullableValuesSpec) { toContainInAnyOrderOnlyValuesFun ->
+                expect {
+                    expect(fluentEmpty()).toContainInAnyOrderOnlyValuesFun(1.0))
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]:
```suggestion
                    expect(fluentEmpty()).toContainInAnyOrderOnlyValuesFun(1.0, emptyArray(), emptyInAnyOrderOnlyReportOptions)
```

The current version (on main) has a helper function defined in nonNullableCases block:
```
        fun Expect<Iterable<Double>>.toContainFun(
            t: Double,
            vararg tX: Double,
            report: InAnyOrderOnlyReportingOptions.() -> Unit = emptyInAnyOrderOnlyReportOptions
        ) = toContainValuesFunArr(t, tX.toTypedArray(), report)
```

That's why in the current version you can use toContainFun(1.0) without the need to pass an array and some options

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    private val toContainInAnyOrderOnlyNullableValuesSpec: Fun3<Iterable<Double?>, Double?, Array<out Double?>, InAnyOrderOnlyReportingOptions.() -> Unit>
+) : ExpectationFunctionBaseTest() {
+
+    @TestFactory
+    fun subjectLessTest() = subjectLessTestFactory(
+        SubjectLessTestData(
+            toContainInAnyOrderOnlyValuesSpec.forSubjectLessTest(
+                2.5,
+                arrayOf(),
+                emptyInAnyOrderOnlyReportOptions)
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: can you please hit CTLR+ALT+L to format the file by intellij

**Review by robstoll — COMMENTED** [review]
let me know if you want more hints and pointers

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt (hunk 2 above)
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+                2.5,
+                arrayOf(),
+                emptyInAnyOrderOnlyReportOptions
+            )
+        )
+    )
+
+    @TestFactory
+    fun empty_collection__1__throws() {
+        nonNullableCases(
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: this won't execute any tests as you don't return your test factory: (same same for other methods). Please execute IterableToContainInAnyOrderOnlyValuesExpectationsTest in the end to verify that all methods are executed as expected: 
```suggestion
    fun empty_collection__1__throws() = nonNullableCases(
```
Hit CTRL+ALT+L (format code) after this change

**Review by robstoll — CHANGES_REQUESTED** [review]
There are a few things left to do (see api-fluent section in the issue description):
- ambiguityTest


Could you please cleanup your commits, in the end I would like to see two commits only. So please squash the latest down to 5af71dfe and the first two commits

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+                    toContainSize(0, 1)
+                }
+            }
+        }
+    }
+
+
+    @TestFactory
+    fun empty_collection__1__and__4__throws() {
+        nonNullableCases(
```
[the remark this answers is not in the document]
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+                        "$successfulBulletPoint$anElementWhichEquals: 2.0",
+                        "$successfulBulletPoint$anElementWhichEquals: 3.0",
+                        "$failingBulletPoint$anElementWhichEquals: 5.0"
+                    )
+                    toContain.exactly(2).value("$successfulBulletPoint$anElementWhichEquals: 4.0")
+                    toContainSize(5, 6)
+                    notToContain(additionalElements, mismatches, mismatchesAdditionalElements)
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+                        "$successfulBulletPoint$anElementWhichEquals: 10.0",
+                        "$successfulBulletPoint$anElementWhichEquals: 11.0",
+                        "$warningBulletPoint$mismatches:",
+                        "${listBulletPoint}5.0",
+                        "${listBulletPoint}8.0"
+                    )
+                    notToContain(additionalElements, mismatchesAdditionalElements)
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+                expect(oneToFour()).toContainFun(
+                    it.first(), it.drop(1).toTypedArray(), emptyInAnyOrderOnlyReportOptions
+                )
+            }
+        }
+    }
+
+    @TestFactory
+    fun additional_elements_and_mismatches() = testFactory(toContainInAnyOrderOnlyValuesSpec) { toContainFun ->
+        expect {
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: you miss the `it(...)` without it you don't define a test

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+                    "$successfulBulletPoint$anElementWhichEquals: 4.0",
+                    "$warningBulletPoint$additionalElements:",
+                    "${listBulletPoint}2.0",
+                    "${listBulletPoint}3.0",
+                    "${listBulletPoint}4.0"
+                )
+                toContainSize(5, 2)
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Oversight by me. Corrected.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+                    toContain.exactly(2).value("$successfulBulletPoint$anElementWhichEquals: 4.0")
+                    toContainSize(5, 6)
+                    notToContain(additionalElements, mismatches, mismatchesAdditionalElements)
+                }
+            }
+        }
+    }
+
+    @TestFactory
+    fun only_failing_with_report_options() = testFactory(toContainInAnyOrderOnlyValuesSpec) { toContainFun ->
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]:
```suggestion
    fun report_options() = testFactory(toContainInAnyOrderOnlyValuesSpec) { toContainFun ->
```

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+                        mismatchesAdditionalElements
+                    )
+                }
+            }
+        }
+    }
+
+    @TestFactory
+    fun shows_all_with_report_option_showAlwaysSummary() =
+        testFactory(toContainInAnyOrderOnlyValuesSpec) { toContainFun ->
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]:
```suggestion
        it("shows all with report option `showAlwaysSummary`") {
```

**Comment by TomerPacific** [comment]
@robstoll - Apologies for the long back and forth. 
Will get on those last comments and migrating the two test files.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt (hunk 9 above)
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

**Comment by TomerPacific** [comment]
@robstoll - I'm looking for the missing it clause and can't find it. I also compared my changes with the original file and saw that in the fused methods, I copied all the it clauses.

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt (hunk 10 above)
Comment (TomerPacific) [reply misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/AbstractIterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

File: apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 11]
```diff
@@ -43,21 +44,21 @@ class IterableToContainInAnyOrderOnlyValuesExpectationsSpec :
...
 
     @Suppress("unused", "UNUSED_VALUE")
-    private fun ambiguityTest() {
-        var list: Expect<List<Number>> = notImplemented()
-        var nList: Expect<Set<Number?>> = notImplemented()
-        var subList: Expect<ArrayList<Number>> = notImplemented()
-        var star: Expect<Collection<*>> = notImplemented()
+    @Test
+    fun ambiguityTest() {
+        var list: Expect<List<Number>> = expect(listOf(1))
```
Comment (robstoll) [inline apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: that's exactly what you need to do, just make sure the test pass in the end

File: apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 12]
```diff
@@ -43,21 +44,21 @@ class IterableToContainInAnyOrderOnlyValuesExpectationsSpec :
...
         list = list toContain o inAny order but only value(1)
         nList = nList toContain o inAny order but only value(1)
         subList = subList toContain o inAny order but only value(1)
         star = star toContain o inAny order but only value(1)
 
-        //TODO type parameter should not be necessary, check with Kotlin 1.4
-        list = list toContain o inAny order but only the values<Number>(1, 1.2)
-        nList = nList toContain o inAny order but only the values<Number>(1, 1.2)
-        subList = subList toContain o inAny order but only the values<Number>(1, 2.2)
+        list = list toContain o inAny order but only the values(1, 1.2)
```
Comment (robstoll) [inline apis/infix/atrium-api-infix/src/commonTest/kotlin/ch/tutteli/atrium/api/infix/en_GB/IterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: e.g. this one will fail since 1.2 is not in there. You need to add a `var listValues = ... expect(listOf(1, 1.2)` nList=... etc. so that you can cover both `value` and `values`

File: apis/fluent/atrium-api-fluent/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/IterableToContainInAnyOrderOnlyValuesExpectationsTest.kt [hunk 13]
```diff
@@ -37,7 +39,8 @@ class IterableToContainInAnyOrderOnlyValuesExpectationsSpec :
     }
 
     @Suppress("unused", "UNUSED_VALUE")
```
[the remark this answers is not in the document]
Comment (TomerPacific) [reply apis/fluent/atrium-api-fluent/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/IterableToContainInAnyOrderOnlyValuesExpectationsTest.kt]: @robstoll - Done.

**Comment by robstoll** [comment]
@TomerPacific thanks for your first contribution. 🙂👍 I'll take over from here

**Comment by robstoll** [comment]
Verify it with `./gradlew -x jsNodeTest build` as there are JS issues (which we discovered now because tests are now also run for JS which was not done when using spek, that's already great 👍). Yes please squash your commits.
